Support shard promotion with Segment Replication.

mch2 opened the pull request:
Signed-off-by: Marc Handalian <handalm@amazon.com>

### Description
This change adds basic failover support with segment replication.  Once selected, a replica will commit its SegmentInfos and reopen a writeable engine.  The replica will also remove all other commits so that this commit is selected when the writeable engine is opened.  It is possible that this commit is not considered 'safe' by the primary, meaning its max seqNo is higher than the global cp.  While an edge case, we never want replicas to reindex with segment replication enabled, so if the global cp has not been updated yet we do not want to revert to a safe commit. This change also updates how SegmentReplicationCheckpointPublisher is wired up within IndexShard so that once promoted the new primary can publish checkpoints.

This PR does not handle edge cases of promotion while there are ongoing replication events, that will be covered in a separate issue.
 
### Issues Resolved
closes #3989
 
### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

dreamer-89 commented on server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationIT.java:
```diff
@@ -70,6 +74,87 @@ protected boolean addMockInternalEngine() {
+        assertSegmentStats(REPLICA_COUNT);
+    }
+
+    private DiscoveryNode getNodeContainingPrimaryShard() {
```
In tests, a index can be created only with one node which means primary will be assigned on that node (replica will be unassinged obviously) and then start another node, which will assign the replica shard on this second node.

mch2 replied:
I've updated the test to follow the new sequence of startup steps, but I've left this for validation so I can assert the replica was promoted correctly.

dreamer-89 commented on server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationIT.java:
```diff
@@ -70,6 +74,87 @@ protected boolean addMockInternalEngine() {
     }
 
+    public void testPrimaryStopped_ReplicaPromoted() throws Exception {
+        final String nodeA = internalCluster().startNode();
+        final String nodeB = internalCluster().startNode();
+        createIndex(INDEX_NAME);
+        ensureGreen(INDEX_NAME);
+
+        client().prepareIndex(INDEX_NAME).setId("1").setSource("foo", "bar").setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE).get();
+        refresh(INDEX_NAME);
+
+        waitForReplicaUpdate();
+        assertHitCount(client(nodeA).prepareSearch(INDEX_NAME).setSize(0).setPreference("_only_local").get(), 1);
+        assertHitCount(client(nodeB).prepareSearch(INDEX_NAME).setSize(0).setPreference("_only_local").get(), 1);
+
+        final DiscoveryNode primaryDiscoveryNode = getNodeContainingPrimaryShard();
```
nit: If `primaryDiscoveryNode` needs to identified, we can start with only one node. This will prevent another call to identify the primary shard node.

```
final String nodeA = internalCluster().startNode(); --> node with primary shard
        createIndex(INDEX_NAME);
        ensureGreen(INDEX_NAME);

// nodeA contains the primary shard, now start second node which will contain the replica shard. Going forward, we can use nodeA as node containing primary shards.
final String nodeB = internalCluster().startNode();
```

Applicable for this and remaining tests.

mch2 commented:
I've added a commit to this ensuring cancelling primary allocation succeeds and that the replica is promoted & primary recreated as a replica.  In testing that I found we were failing to publish a replication checkpoint if the primary flushed during close. That is now fixed, the shard must be open for us to publish the replication cp.

Bukhtawar commented on server/src/main/java/org/opensearch/index/shard/IndexShard.java:
```diff
@@ -4111,4 +4110,26 @@ RetentionLeaseSyncer getRetentionLeaseSyncer() {
+        getReplicationEngine().ifPresentOrElse(engine -> {
+            try {
+                engine.commitSegmentInfos();
+                resetEngineToGlobalCheckpoint();
```
1. Can we add checks/assertions on currentGlobalCheckpoint and maxSeqNo before resetting engine to Global checkpoint
2. Can we add assertions on current routing and new routing as well

mch2 replied:
1.  With segrep enabled we don't want to reindex ops that have already been indexed.  To avoid this the replica does not actually reset to the global cp.  In this method, when we commit we remove all other commits so the new one is always selected when opening InternalEngine. This means we allow the local cp and/or max seqNo in the commit to be higher than the global cp and not actually "safe", but I don't think we care? The use of `resetEngineToGlobalCheckpoint` here is reused to swap the engine type, and potentially replay up to the global cp.

2. Added this before this method is invoked inside `updateShardState`.

Bukhtawar commented on server/src/main/java/org/opensearch/index/shard/IndexShard.java:
```diff
@@ -4111,4 +4110,26 @@ RetentionLeaseSyncer getRetentionLeaseSyncer() {
+            try {
+                engine.commitSegmentInfos();
+                resetEngineToGlobalCheckpoint();
+            } catch (IOException e) {
+                throw new OpenSearchException("Unable to change engine type, failing", e);
+            }
+        }, () -> { throw new OpenSearchException("Expected replica engine to be of type NRTReplicationEngine"); });
+    }
```
Lets add more details around shardId

mch2 commented on server/src/internalClusterTest/java/org/opensearch/indices/replication/SegmentReplicationIT.java:
```diff
@@ -303,4 +388,25 @@ private Map<String, Segment> getLatestSegments(ShardSegments segments) {
     }
+
+    @Nullable
+    private ShardRouting getShardRoutingForNodeName(String nodeName) {
+        final ClusterState state = client(internalCluster().getClusterManagerName()).admin().cluster().prepareState().get().getState();
+        for (IndexShardRoutingTable shardRoutingTable : state.routingTable().index(INDEX_NAME)) {
+            for (ShardRouting shardRouting : shardRoutingTable.activeShards()) {
+                final String nodeId = shardRouting.currentNodeId();
+                final DiscoveryNode discoveryNode = state.nodes().resolveNode(nodeId);
+                if (discoveryNode.getName().equals(nodeName)) {
+                    return shardRouting;
+                }
+            }
+        }
+        return null;
```
thanks!
